packages/teleterm/src/ui/DocumentCluster/ClusterResources/useServerSideResources.ts

```diff
@@ -0,0 +1,153 @@
+import { useState, useEffect } from 'react';
+import { SortType } from 'design/DataTable/types';
+import { useAsync } from 'shared/hooks/useAsync';
+import { ServerSideParams } from 'teleterm/services/tshd/types';
+import { useAppContext } from 'teleterm/ui/appContextProvider';
+import { retryWithRelogin } from 'teleterm/ui/utils';
+import { useClusterContext } from '../clusterContext';
+import { AgentFilter, AgentLabel } from 'teleport/services/agents';
+import { addAgentLabelToQuery } from 'shared/utils/addAgentLabelToQuery';
+import { ResourceKind } from 'teleport/Discover/Shared';
+
+export function useServerSideResources(resource: ResourceKind) {
+  const ctx = useAppContext();
+  const { clusterUri, documentUri } = useClusterContext();
+  const [pageIndex, setPageIndex] = useState(0);
+  const [keys, setKeys] = useState<string[]>([]);
+  const [agentFilter, setAgentFilter] = useState<AgentFilter>({
+    sort: getDefaultSort(),
+  });
+  const [pageCount, setPageCount] = useState({
+    to: 0,
+    from: 0,
+    total: 0,
+  });
+
+  const limit = 15;
+
+  function getFetchCallback(params: ServerSideParams) {
+    switch (resource) {
+      case ResourceKind.Server:
+        return retryWithRelogin(ctx, documentUri, clusterUri, () =>
+          ctx.clustersService.fetchServers(params)
+        );
+      default: {
+        throw new Error(`Fetch not implemented for: ${resource}`);
+      }
+    }
+  }
```

Comment:
Originally I was planning to just send in the clustersService function to the call of this function in `useServers` but was getting some big typescript errors. If I typed the callback going into `useServers` I would get errors that `useServers` didn't know what type was coming back from the `fetchAttempt`, and if I typed it, then `retryWithRelogin` was throwing typescript errors of `Cannot return Promise<unknown>` or something along those lines. 

This looks a bit ugly but, it pleased the typescript gods and isn't too bad imo. 

Ideally I would have it the first suggested way of `useServers(ctx.clustersService.fetchServers)` but I hit my limit of typescript arcane knowledge I think.

Secondly, we couldn't have the `retryWithRelogin` wrapper outside of this file (example like moving it to `useServers(retryWithRelogin....)` because the `params` needed for the internal callback are held within this file instead.

If i've missed something simple let me know and I'm happy to update!

Reply:
The attached patch makes passing the function work as expected. I used generic type to extract the parts of the response that change between calls to `useServerSideResources`.

Unfortunately, we cannot simply do `useServerSideResources(appContext.clustersService.fetchServers)` because in JS instance methods are not bound to the instance. We'd have to do

```typescript
useServerSideResources(appContext.clustersService.fetchServers.bind(appContext.clustersService))
```

but Function.bind loses all type information so we'd also need to add

```typescript
useServerSideResources<types.Server>(appContext.clustersService.fetchServers.bind(appContext.clustersService))
```

at which point it's no longer worth it I think.

<details>
<summary>Patch</summary>

```diff
diff --git a/packages/teleterm/src/ui/DocumentCluster/ClusterResources/Servers/useServers.ts b/packages/teleterm/src/ui/DocumentCluster/ClusterResources/Servers/useServers.ts
index 45bfe80d..4b29c15c 100644
--- a/packages/teleterm/src/ui/DocumentCluster/ClusterResources/Servers/useServers.ts
+++ b/packages/teleterm/src/ui/DocumentCluster/ClusterResources/Servers/useServers.ts
@@ -13,8 +13,8 @@ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
 */
-import { ResourceKind } from 'teleport/Discover/Shared';
 import { useAppContext } from 'teleterm/ui/appContextProvider';
+
 import { useServerSideResources } from '../useServerSideResources';
 
 export function useServers() {
@@ -30,7 +30,9 @@ export function useServers() {
     onAgentLabelClick,
     updateSort,
     pageCount,
-  } = useServerSideResources(ResourceKind.Server);
+  } = useServerSideResources(params =>
+    appContext.clustersService.fetchServers(params)
+  );
 
   function getSshLogins(serverUri: string): string[] {
     const cluster = appContext.clustersService.findClusterByResource(serverUri);
diff --git a/packages/teleterm/src/ui/DocumentCluster/ClusterResources/useServerSideResources.ts b/packages/teleterm/src/ui/DocumentCluster/ClusterResources/useServerSideResources.ts
index 30c535f1..6fb55a89 100644
--- a/packages/teleterm/src/ui/DocumentCluster/ClusterResources/useServerSideResources.ts
+++ b/packages/teleterm/src/ui/DocumentCluster/ClusterResources/useServerSideResources.ts
@@ -1,15 +1,21 @@
 import { useState, useEffect } from 'react';
+
 import { SortType } from 'design/DataTable/types';
 import { useAsync } from 'shared/hooks/useAsync';
+import { AgentFilter, AgentLabel } from 'teleport/services/agents';
+import { addAgentLabelToQuery } from 'shared/utils/addAgentLabelToQuery';
+
 import { ServerSideParams } from 'teleterm/services/tshd/types';
 import { useAppContext } from 'teleterm/ui/appContextProvider';
 import { retryWithRelogin } from 'teleterm/ui/utils';
+
 import { useClusterContext } from '../clusterContext';
-import { AgentFilter, AgentLabel } from 'teleport/services/agents';
-import { addAgentLabelToQuery } from 'shared/utils/addAgentLabelToQuery';
-import { ResourceKind } from 'teleport/Discover/Shared';
 
-export function useServerSideResources(resource: ResourceKind) {
+export function useServerSideResources<T>(
+  fetchFunction: (
+    params: ServerSideParams
+  ) => Promise<{ agentsList: Array<T>; totalCount: number; startKey: string }>
+) {
   const ctx = useAppContext();
   const { clusterUri, documentUri } = useClusterContext();
   const [pageIndex, setPageIndex] = useState(0);
@@ -25,28 +31,18 @@ export function useServerSideResources(resource: ResourceKind) {
 
   const limit = 15;
 
-  function getFetchCallback(params: ServerSideParams) {
-    switch (resource) {
-      case ResourceKind.Server:
-        return retryWithRelogin(ctx, documentUri, clusterUri, () =>
-          ctx.clustersService.fetchServers(params)
-        );
-      default: {
-        throw new Error(`Fetch not implemented for: ${resource}`);
-      }
-    }
-  }
-
   // startKey is used here as a way to paginate through agents returned from
   // their respective rpcs.
-  const [fetchAttempt, fetch] = useAsync(async (startKey: string) => {
-    return getFetchCallback({
-      ...agentFilter,
-      limit,
-      clusterUri,
-      startKey,
-    });
-  });
+  const [fetchAttempt, fetch] = useAsync(async (startKey: string) =>
+    retryWithRelogin(ctx, documentUri, clusterUri, () =>
+      fetchFunction({
+        ...agentFilter,
+        limit,
+        clusterUri,
+        startKey,
+      })
+    )
+  );
 
   // If there is no startKey at the current page's index, there is no more data to get
   function nextPage() {
```
</details>

Reply:
You might want to use a less vague name for the generic type instead of `T`. `Agent` would be good.